python/sqlflow_playground/playground_server_design.md

```diff
@@ -0,0 +1,135 @@
+```
+This service always uses `HTTPS` and only accepts authorized clients
+by checking their certification file. So there is no dedicated api
+for user authentication.
+
+Currently supported API are:
+| name | method | params | description |
+| - | - | - | - |
+| create_db | POST | {"user_id": "id"} | create a DB for given user, json param |
+| heart_beat| GET  | user_id=id | report a heart beat of given client |
+
+
+## How to Use
+
+### For Service Maintainer
+The maintainer should [provide the playground cluster](../dev.md), and
+bootup a `SQLFlow Playground Server`.  The server should have privillege
+to access the `kubectl` command of the cluster.  To install the server,
+maintainer can use below command:
+```bash
+    mkdir $HOME/workspace
+    cd $HOME/workspace
+    pip install sqlflow_playground
+    mkdir key_store
+    gen_cert.sh server
+    sqlflow_playground --port=50052 \
+      --ca_crt=key_store/ca/ca.crt \
+      --server_key=key_store/server/server.key \
+      --server_crt=key_store/server/server.crt
+```
+In the above commands, we first installed the sqlflow playground package
+which carries the main cluster operation logic.  Then, we use the key
+tool to generate a server certification file (Of course, it's not necessary
+if you have your own certification files) which enables us to provide
+`HTTPS` service.  Finally, we start the `REST API` service at port 50052.
+
+Our playground service uses bi-directional validation.  So, the maintainer
+needs to generate a certification file for a trusted user. Use below command and
+send the generated `.crt` and `.key` file together with the `ca.crt` to
+the user.
+
+```bash
+    gen_cert.sh some_client
+```
+
+### For The User
+
+To use this service, the user should get authorized from the playground's maintainer.
+In detail, user should get `ca.crt`, `client.key` and the `client.crt` file from
+the maintainer and keep them in some very-safe place. Also, the user should ask
+the maintainer for the sqlflow server address and the sqlflow playground server
+address. Then, the user will install Jupyter Notebook and the SQLFlow plugin package
+and do some configuration. Finally, the user can experience SQLFlow in his Jupyter 
+Notebook.
+
+```bash
+    pip3 install notebook sqlflow==0.15.0
+
+    cat >$HOME/.sqlflow_playground.env <<EOF
+SQLFLOW_SERVER="{sqlflow server address}"
+SQLFLOW_PLAYGROUND_USER_ID_ENV=SQLFLOW_USER_ID
+SQLFLOW_USER_ID="{your name}"
+SQLFLOW_PLAYGROUND_SERVRE="{sqlflow playground server address}"
+SQLFLOW_PLAYGROUND_SERVER_CA="{path to your ca.crt file}"
+SQLFLOW_PLAYGROUND_CLIENT_KEY="{path to your client.key file}"
+SQLFLOW_PLAYGROUND_CLIENT_CERT="{path to your client.crt file}"
+EOF
+
+    export SQLFLOW_JUPYTER_ENV_PATH="$HOME/.sqlflow_playground.env"
+    # start the notebook and try use %%sqlflow magic command
+    jupyter notebook
+```
+
+## Implementation
+
+We use [tornado](https://www.tornadoweb.org/) as the web framework which provides
```

Comment:
When does the plugin tell the playground server to create a new user?

Reply:
It's in this PR https://github.com/sql-machine-learning/pysqlflow/pull/110, when we first use sqlflow magic command, we try to create a new user and ask for some resource.